File: java/datastructure/linkedlist/leetcode/LinkedListCycle.java
package datastructure.linkedlist.leetcode;

import java.util.HashSet;
import java.util.Set;

/**
 * @author roseduan
 * @time 2020/10/24 2:25 下午
 * @description 环形链表
 */
public class LinkedListCycle {

    /**
     * 使用一个set判重
     */
    public boolean hasCycle1(ListNode head) {
        Set<ListNode> set = new HashSet<>();
        while (head != null) {
            if (set.contains(head)) {
                return true;
            }
            set.add(head);
            head = head.next;
        }
        return false;
    }

    /**
     * 快慢指针法
     */
    public boolean hasCycle2(ListNode head) {
        ListNode slow = head, fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) {
                return true;
            }
        }
        return false;
    }

    public static class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
    }
}

File: cpp/linkedlist/linked_list_cycle.cpp
// 检测链表是否有环
#include "unordered_set"
using namespace std;

struct ListNode {
    int val;
    ListNode *next;

    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:

    // 使用一个哈希表
    bool hasCycle1(ListNode *head) {
        unordered_set<ListNode*> s;
        while (head) {
            auto it = s.find(head);
            if (it != s.end()) {
                return true;
            }
            head = head->next;
        }
        return false;
    }

    // 快慢指针
    bool hasCycle2(ListNode *head) {
        ListNode* slow = head;
        ListNode* fast = head;
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
            if (slow == fast) {
                return true;
            }
        }
        return false;
    }
};
